lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -2223,6 +2292,71 @@ instance {-# OVERLAPS #-} EncodeStakeAddress n
   where
     toJSON (acct, _) = toJSON . encodeStakeAddress @n . getApiT $ acct
 
+instance FromJSON ApiSharedWalletPostDataFromAccountPubX where
+    parseJSON = genericParseJSON defaultRecordTypeOptions
+instance ToJSON ApiSharedWalletPostDataFromAccountPubX where
+    toJSON = genericToJSON defaultRecordTypeOptions
+
+instance FromJSON ApiSharedWalletPostDataFromMnemonics where
+    parseJSON = genericParseJSON defaultRecordTypeOptions
+instance ToJSON ApiSharedWalletPostDataFromMnemonics where
+    toJSON = genericToJSON defaultRecordTypeOptions
+
+instance FromJSON ApiSharedWalletPostData where
+    parseJSON obj = do
+        mnemonic <-
+            (withObject "postData" $
+             \o -> o .:? "mnemonic_sentence" :: Aeson.Parser (Maybe [Text])) obj
+        case mnemonic of
+            Nothing -> do
+                xs <- parseJSON obj :: Aeson.Parser ApiSharedWalletPostDataFromAccountPubX
+                pure $ ApiSharedWalletPostData $ Right xs
+            _ -> do
+                xs <- parseJSON obj :: Aeson.Parser ApiSharedWalletPostDataFromMnemonics
+                pure $ ApiSharedWalletPostData $ Left xs
+
+instance ToJSON ApiSharedWalletPostData where
+    toJSON (ApiSharedWalletPostData (Left c))= toJSON c
```

Comment:
When there are nested record fields as suggested above, then a function to merge toJSON results will help:
```haskell
combine :: Value -> Value -> Value
combine a b = case (a, b) of
    (Object oa, Object ob) -> Json.Object (oa <> ob)
    (_, otherValue) -> otherValue  -- don't use this
```